File: Go/XClient/config_test.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XClient

import (
	"reflect"
	"testing"

	"github.com/frameworkease/Universal.X.gRPC/Go/internal/Proto/Greet"
	"github.com/stretchr/testify/assert"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

func TestConfigure(t *testing.T) {
	defer func() {
		anyType := reflect.TypeFor[any]()
		fooType := reflect.TypeFor[Greet.FooServiceClient]()

		configsMu.Lock()
		delete(configs, anyType)
		delete(configs, fooType)
		configsMu.Unlock()
	}()

	gateway := "http://127.0.0.1:50051"
	options := grpc.WithTransportCredentials(insecure.NewCredentials())

	tests := []struct {
		name      string
		configure func(string, grpc.DialOption)
		typ       reflect.Type
	}{
		{
			name:      "Shared",
			configure: func(g string, o grpc.DialOption) { Configure[any](g, o) },
			typ:       reflect.TypeFor[any](),
		},
		{
			name:      "Generic",
			configure: func(g string, o grpc.DialOption) { Configure[Greet.FooServiceClient](g, o) },
			typ:       reflect.TypeFor[Greet.FooServiceClient](),
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			defer func() {
				configsMu.Lock()
				delete(configs, tt.typ)
				configsMu.Unlock()
			}()
			tt.configure(gateway, options)

			loadedConfig, ok := configs[tt.typ]
			assert.True(t, ok, "配置应当存在。")
			assert.Equal(t, gateway, loadedConfig.gateway, "网关地址应当匹配。")
			assert.Equal(t, options, loadedConfig.options[0], "通道选项应当匹配。")
		})
	}
}

File: CSharp/tests~/Assets/Tests/Runtime/Scripts/XClient/config_test.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using FrameworkEase.Universal.X.gRPC;
using FrameworkEase.Universal.X.gRPC.Tests.Proto.Greet;
using Grpc.Net.Client;
using NUnit.Framework;

public partial class TestXClient
{
    [Test]
    public void Configure()
    {
        var gateway = "http://127.0.0.1:50051";
        var options = new GrpcChannelOptions
        {
#if UNITY_5_3_OR_NEWER
            HttpHandler = new Cysharp.Net.Http.YetAnotherHttpHandler() { Http2Only = true }
#else
            HttpHandler = new System.Net.Http.SocketsHttpHandler { EnableMultipleHttp2Connections = true }
#endif
        };

        #region Shared
        {
            XClient.Configure(gateway, options);

            Assert.That(XClient.Configuration.Gateway, Is.EqualTo(gateway), "共享 Gateway 应当被设置。");
            Assert.That(XClient.Configuration.Options, Is.EqualTo(options), "共享 Options 应当被设置。");

            XClient.Configuration.Gateway = null;
            XClient.Configuration.Options = null;
        }
        #endregion

        #region Generic
        {
            XClient<FooService.FooServiceClient>.Configure(gateway, options);

            Assert.That(XClient<FooService.FooServiceClient>.Configuration.Gateway, Is.EqualTo(gateway), "专用 Gateway 应当被设置。");
            Assert.That(XClient<FooService.FooServiceClient>.Configuration.Options, Is.EqualTo(options), "专用 Options 应当被设置。");

            XClient<FooService.FooServiceClient>.Configuration.Gateway = null;
            XClient<FooService.FooServiceClient>.Configuration.Options = null;
        }
        #endregion
    }
}
